common/server_player.cpp

```diff
@@ -994,27 +992,24 @@ Server_Player::cmdMulligan(const Command_Mulligan & /*cmd*/, ResponseContainer &
     }
 
     Server_CardZone *hand = zones.value("hand");
-    int number = (hand->getCards().size() <= 1) ? initialCards : hand->getCards().size() - 1;
-
     Server_CardZone *deck = zones.value("deck");
-    while (!hand->getCards().isEmpty()) {
+    int number = cmd.number();
+
+    auto cardsToMove = QList<const CardToMove *>();
+    for (auto &card : hand->getCards()) {
         auto *cardToMove = new CardToMove;
-        cardToMove->set_card_id(hand->getCards().first()->getId());
-        moveCard(ges, hand, QList<const CardToMove *>() << cardToMove, deck, 0, 0, false);
-        delete cardToMove;
+        cardToMove->set_card_id(card->getId());
+        cardsToMove.append(cardToMove);
     }
+    moveCard(ges, hand, cardsToMove, deck, -1, 0, false);
```

Comment:
Seems like `cardsToMove` leaks memory, qDeleteAll()?

Reply:
I was worried about that, seems like I forgot about it before pushing.